compiler/rustc_mir_build/src/build/expr/as_place.rs

```diff
@@ -20,13 +28,132 @@ use rustc_index::vec::Idx;
+        ty::Closure(_, substs) => ty::UpvarSubsts::Closure(substs),
+        ty::Generator(_, substs, _) => ty::UpvarSubsts::Generator(substs),
+        _ => bug!("Lowering capture for non-closure type {:?}", closure_ty),
+    };
+
+    // Access the capture by accessing the field within the Closure struct.
+    //
+    // We must have inferred the capture types since we are building MIR, therefore
+    // it's safe to call `upvar_tys` and we can unwrap here because
+    // we know that the capture exists and is the `capture_index`-th capture.
+    let var_ty = substs.upvar_tys().nth(capture_index).unwrap();
+
+    let upvar_id = ty::UpvarId::new(var_hir_id, closure_def_id.expect_local());
+    let capture_kind = typeck_results.upvar_capture(upvar_id);
+
+    Some((capture_index, var_ty, capture_kind))
+}
+
+/// Takes a PlaceBuilder and resolves the upvar (if any) within it,
+/// so that the PlaceBuilder now starts from PlaceBase::Local.
+///
+/// Returns a Result with the error being the HirId of the
+/// Upvar that was not found.
+fn to_upvars_resolved_place_builder<'a, 'tcx>(
+    from_builder: PlaceBuilder<'tcx>,
+    tcx: TyCtxt<'tcx>,
+    typeck_results: &'a ty::TypeckResults<'tcx>,
+) -> Result<PlaceBuilder<'tcx>, HirId> {
+    match from_builder.base {
+        PlaceBase::Local(_) => Ok(from_builder),
+        PlaceBase::Upvar { var_hir_id, closure_def_id, closure_kind } => {
+            // Captures are represented using fields inside a structure.
+            // This represents accessing self in the closure structure
+            let local = PlaceBase::Local(Local::new(1));
+            let mut place_builder = PlaceBuilder::from(local);
+            match closure_kind {
+                ty::ClosureKind::Fn | ty::ClosureKind::FnMut => {
+                    place_builder = place_builder.deref();
+                }
+                ty::ClosureKind::FnOnce => {}
+            }
+
+            let (capture_index, var_ty, capture_kind) =
+                if let Some(capture_details) = capture_matching_projections(
+                    tcx,
+                    typeck_results,
+                    var_hir_id,
+                    closure_def_id,
+                    &from_builder.projection,
+                ) {
+                    capture_details
+                } else {
+                    return Err(var_hir_id);
+                };
+
+            place_builder = place_builder.field(Field::new(capture_index), var_ty);
+
+            // If the variable is captured via ByRef(Immutable/Mutable) Borrow,
+            // we need to deref it
+            place_builder = match capture_kind {
+                ty::UpvarCapture::ByRef(_) => place_builder.deref(),
+                ty::UpvarCapture::ByValue(_) => place_builder,
+            };
+
+            let next_projection = 0;
+            let mut curr_projections = from_builder.projection;
+            place_builder.projection.extend(curr_projections.drain(next_projection..));
```

Comment:
`place_builder.projection.extend(curr_projections);`?

Reply:
I did this on purpose so that next commit we can update next_projection to be `captured_place.projections.len()`